utils/loading_shared_values.hh

```diff
@@ -260,6 +260,30 @@ public:
         }
     }
 
+    /// \brief
+    /// Returns an entry_ptr for the given key. If the value is already present, it is returned.
+    /// If the entry is loading, or it is missing, the parameter is used as the value.
+    entry_ptr get_or_insert(const key_type& key, value_type value) {
```

Comment:
I don't understand this get_or_insert(). If you already have the value, why bother the cache? Just use the value you have.

Reply:
This method is not actually for getting the value, but for inserting it into the container. Normally I'd introduce something like just insert(), but here the entries are alive only as long as there are external references to them, so the insert needs to also get from the shared_loading_values, and if it needs to get() either way, I made it a bit more general, allowing to both insert and get, hence the name.
I also considered making this method for inserting only, it would be named something like insert_and_get(), and it would assert that there is no entry for this key, but I decided it would be even worse than what I ended up with

Reply:
I don't understand it. If they value is already present, why did you spend the effort to calculate it?

The cache should only calculate (get) the value if it's not known.

Reply:
> I don't understand it. If they value is already present, why did you spend the effort to calculate it?
> 
> The cache should only calculate (get) the value if it's not known.

@avikivity @wmitros Please, note that you don't need to add any new API to insert an existing value into the cache - you can use an existing `get_ptr(key, loader)` and pass as a `loader` a lambda that `return std::move()`s the existing value you'd pass it.

Reply:
@avikivity I'm not sure if we're using the same naming here. The thing that we want to avoid is calling the `loader` because that's where we may be calculating a new value (and in this new method we do not call the loader). "Getting" and "inserting" are operations on an underlying unordered_set, so I'd say they're not particularly expensive. 
@vladzcloudius In this PR we're discussing 2 variants of changing values in the cache - either modifying an entry that's in the cache the entire time, or removing and inserting a new entry into the cache. The comment is only relevant for the second option, so I'll answer it in that context. 
It's true that we can use a `loader` that inserts a ready value, the question is: can we assure that when when we call `get_ptr` in multiple continuations, the loader will be selected over the loaders from other `get_ptr`s? `get_ptr` returns a future, so I'd expect any other `get_ptr` that's called after the one with the ready value may get executed before that one
Here this is assured by the fact that when we have a ready value, we put it inside the cache synchronously. Actually, we could change the `put` and `get_or_insert` to something like `get_ptr_sync` and 'get_or_load_sync' where we would do everything synchronously (it's almost what we're doing already)